chain/chain/src/chain.rs

```diff
@@ -3077,6 +3077,39 @@ impl Chain {
         Ok(self.chain_store.get_outcomes_by_id(id)?.into_iter().map(Into::into).collect())
     }
 
+    /// Returns execution status based on the list of currently existing outcomes
+    fn get_execution_status(
+        &self,
+        outcomes: &[ExecutionOutcomeWithIdView],
+    ) -> FinalExecutionStatus {
+        let mut awaiting_receipt_ids = HashSet::new();
+        let mut seen_receipt_ids = HashSet::new();
+        let mut result: FinalExecutionStatus = FinalExecutionStatus::NotStarted;
+        for outcome in outcomes {
+            seen_receipt_ids.insert(&outcome.id);
+            match &outcome.outcome.status {
+                ExecutionStatusView::Unknown => return FinalExecutionStatus::Started,
+                ExecutionStatusView::Failure(e) => return FinalExecutionStatus::Failure(e.clone()),
+                ExecutionStatusView::SuccessValue(v) => {
+                    if result == FinalExecutionStatus::NotStarted {
+                        // historically, we used the first SuccessValue we have seen
+                        // let's continue sticking to it
+                        result = FinalExecutionStatus::SuccessValue(v.clone());
+                    }
+                }
+                ExecutionStatusView::SuccessReceiptId(_) => {
+                    awaiting_receipt_ids.extend(&outcome.outcome.receipt_ids);
+                }
+            }
+        }
+        return if (&awaiting_receipt_ids - &seen_receipt_ids).is_empty() {
+            result
+        } else {
+            FinalExecutionStatus::Started
+        };
+    }
```

Comment:
I've decided to extract this logic from `get_final_transaction_result` to a separate function, and rewrite it, fixing a number of potential bugs in the meantime.

Some important context:
The receipt may produce tons of new receipts. However, the corresponding execution outcome will have the status `ExecutionStatusView::SuccessReceiptId` with only one id inside - it's the last receipt id of all the produced receipts. So if you want to see the status, you have to go through all the list, it's not enough to look at the status of the id inside `SuccessReceiptId` field.

-----

Results

- I stopped relying on the order of the receipts
- I started checking all the produced receipts statuses, not only the last one
- I fixed a bug. Before, we have
```
match &outcome_with_id.outcome.status {
                         ExecutionStatusView::Unknown if num_outcomes == 1 => {
                             Some(FinalExecutionStatus::NotStarted)
                         }
                         ...
}

```
In reality, if we have 1 outcome, it means the status is at least Started.